### PR TITLE
[Bugfix][FP8 linear oracle] Verify scale dtype (e8m0 vs fp32) compatibility in FP8 linear backends

### PR DESCRIPTION
**Pull request by fxmarty-amd**
## Purpose

Target model e.g. http://huggingface.co/deepseek-ai/DeepSeek-V4-Flash

Fix FP8 block-linear kernel selection for E8M0 weight scales. The FP8 and Quark W8A8 FP8 paths now advertise E8M0 scale storage through `kFp8Static128BlockE8M0Sym`, so selection can reject backends that cannot consume it.

This specifically fixes incorrect E8M0 handling in:
- `HummingFP8ScaledMMLinearKernel`: currently wrong `self.config.weight_quant_key.scale.dtype`: https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/vllm/model_executor/kernels/linear/scaled_mm/humming.py#L50-L60
- `CutlassFp8BlockScaledMMKernel`: CUTLASS consumes block scales as float32 and must reject E8M0 storage: https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/csrc/libtorch_stable/quantization/w8a8/cutlass/c3x/scaled_mm_blockwise_sm90_fp8_dispatch.cuh#L56. The SM90 callsite is https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/csrc/libtorch_stable/quantization/w8a8/cutlass/c3x/scaled_mm_blockwise_sm90_fp8.cu#L13-L14, which is selected by https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/csrc/libtorch_stable/quantization/w8a8/cutlass/scaled_mm_c3x_sm90.cu#L17-L20. The top-level CUTLASS dispatcher calls the latter at https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/csrc/libtorch_stable/quantization/w8a8/cutlass/scaled_mm_entry.cu#L236-L241. SM100 and SM120 use the same float32 block-scale type: https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/csrc/libtorch_stable/quantization/w8a8/cutlass/c3x/scaled_mm_blockwise_sm100_fp8_dispatch.cuh#L57 and https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/csrc/libtorch_stable/quantization/w8a8/cutlass/c3x/scaled_mm_blockwise_sm120_fp8_dispatch.cuh#L58, respectively.
- `CPUFp8BlockScaledMMKernel`: forwards the block scale directly to the CPU op, which requires float32 scales: https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/vllm/model_executor/kernels/linear/scaled_mm/cpu.py#L300-L316 and https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/csrc/cpu/sgl-kernels/gemm_fp8.cpp#L1150
- `TritonFp8BlockScaledMMKernel`: `accumulator += tl.dot(a, b) * a_s[:, None] * b_s[None, :]` used on CUDA, guarded only on ROCm/XPU: https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/vllm/model_executor/layers/quantization/utils/fp8_utils.py#L920-L927

| Backend | E8M0 support | Evidence |
| --- | --- | --- |
| `DeepGemmFp8BlockScaledMMKernel` | Supported | [Upcasts checkpoint E8M0 scales and packs the DeepGEMM layout](https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/vllm/model_executor/layers/quantization/utils/fp8_utils.py#L1104-L1115). |
| `FlashInferFp8DeepGEMMDynamicBlockScaledKernel` | Supported | [Uses DeepGEMM post-processing for the shared weight-scale layout](https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/vllm/model_executor/kernels/linear/scaled_mm/flashinfer.py#L187-L189). |
| `AiterFp8BlockScaledMMKernel` | Supported | [Upcasts E8M0 scales during weight processing](https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/vllm/model_executor/kernels/linear/scaled_mm/aiter.py#L390-L401). |
| `B12xFp8BlockScaledMMKernel` | Supported | [Upcasts E8M0 scales before its GEMM path](https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/vllm/model_executor/kernels/linear/scaled_mm/b12x.py#L101-L119). |
| `HummingFP8ScaledMMLinearKernel` | Supported | [Maps E8M0 to `torch.float8_e8m0fnu`](https://github.com/inclusionAI/humming/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/humming/dtypes.py#L200-L206). |
| `MarlinFP8ScaledMMLinearKernel` | Supported | [Dispatches FP8 weights with E8M0 scales to the E8M0 scale type](https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/csrc/libtorch_stable/quantization/marlin/marlin.cu#L601-L615). |
| `CutlassFp8BlockScaledMMKernel` | Unsupported | [CUTLASS block kernels consume scale storage as float32](https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/csrc/libtorch_stable/quantization/w8a8/cutlass/c3x/scaled_mm_blockwise_sm90_fp8_dispatch.cuh#L56), so E8M0 scales are rejected during selection. |
| `CPUFp8BlockScaledMMKernel` | Unsupported | [The CPU FP8 op requires float32 scales](https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/csrc/cpu/sgl-kernels/gemm_fp8.cpp#L1150). |
| `XPUFp8BlockScaledMMKernel` | Supported | See https://github.com/vllm-project/vllm-xpu-kernels/pull/235 |
| `BlockWiseTorchFP8ScaledMMLinearKernel` | Unsupported | [Documents float32 block scales and forwards them directly to `torch._scaled_mm`](https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/vllm/model_executor/kernels/linear/scaled_mm/pytorch.py#L261-L270). |
| CUDA `TritonFp8BlockScaledMMKernel` | Unsupported | [States that Triton cannot bind E8M0 scales directly, but only decodes them on ROCm/XPU](https://github.com/vllm-project/vllm/blob/5ee84d3c5207bd59faa53e19e60fa82e34961444/vllm/model_executor/layers/quantization/utils/fp8_utils.py#L920-L927). |

This change applies to `Fp8LinearMethod`, `QuarkW8A8Fp8PerBlock` only as:

- `Fp8PerBlockOnlineLinearMethod`,
- `FBGEMMFp8LinearMethod`,
- `ModelOptFp8PbWoLinearMethod`,
- `CompressedTensorsW8A8Fp8`

do NOT support e8m0 scale dtype.

CUDA 13.0 + H100 rejects e8m0 scales:

```python
import torch

def run(scale_b: torch.Tensor) -> None:
    m, k, n = 4, 128, 128
    a = torch.randn(m, k, device="cuda", dtype=torch.bfloat16).to(torch.float8_e4m3fn)
    b = torch.randn(n, k, device="cuda", dtype=torch.bfloat16).to(
        torch.float8_e4m3fn).t()
    scale_a = torch.ones((m, k // 128), device="cuda", dtype=torch.float32)
    output = torch._scaled_mm(a, b, scale_a=scale_a, scale_b=scale_b, out_dtype=torch.bfloat16)
    if isinstance(output, tuple):
        output = output[0]
    print(f"scale_b={scale_b.dtype}: succeeded, output={output.dtype} {tuple(output.shape)}")

if not torch.cuda.is_available():
    raise RuntimeError("CUDA is required")

print(f"torch={torch.__version__}")
print(f"cuda={torch.version.cuda}")
print(f"device={torch.cuda.get_device_name()}")

run(torch.ones((1, 1), device="cuda", dtype=torch.float32))
run(torch.ones((1, 1), device="cuda", dtype=torch.float8_e8m0fnu))
```

gives

`For BlockWise 128x128 scaling, a and b should be float8, scales should be float, scale_a should be (1, 1) and scale_b should be (1, 1), and both should be near-inner-dim-major (with 16-byte aligned strides).`

## Test Plan

- `pytest tests/kernels/quantization/test_fp8_kernel_selection.py`

## Test Results

`14 passed, 14 warnings in 4.92s`

### CODE REVIEW

File: tests/kernels/quantization/test_fp8_kernel_selection.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+_E8M0_BACKEND_SUPPORT = [
+    (FlashInferFp8DeepGEMMDynamicBlockScaledKernel, "cuda", True),
+    (DeepGemmFp8BlockScaledMMKernel, "cuda", True),
+    (CutlassFp8BlockScaledMMKernel, "cuda", False),
+    (B12xFp8BlockScaledMMKernel, "cuda", True),
+    (MarlinFP8ScaledMMLinearKernel, "cuda", True),
+    (HummingFP8ScaledMMLinearKernel, "cuda", True),
+    (TritonFp8BlockScaledMMKernel, "cuda", False),
+    (BlockWiseTorchFP8ScaledMMLinearKernel, "cuda", False),
```
Comment (tjtanaa) [inline tests/kernels/quantization/test_fp8_kernel_selection.py:63]: We also need to test `BlockWiseTorchFP8ScaledMMLinearKernel` on ROCm, if you check its `can_implement`, it includes ROCm.
`(BlockWiseTorchFP8ScaledMMLinearKernel, "rocm", False),`

Can you ask agent to look through to make sure all the backends that are relevant to each platform are included?

File: vllm/model_executor/kernels/linear/scaled_mm/xpu.py [hunk 2]
```diff
@@ -198,6 +198,18 @@ def is_supported(
...
 
+    @classmethod
+    def can_implement(
+        cls, config: FP8ScaledMMLinearLayerConfig
+    ) -> tuple[bool, str | None]:
+        can_implement_base, reason = super().can_implement(config)
+        if not can_implement_base:
+            return can_implement_base, reason
+
+        if config.weight_quant_key.scale.dtype == torch.float8_e8m0fnu:
```
Comment (tjtanaa) [inline vllm/model_executor/kernels/linear/scaled_mm/xpu.py]: @xuechendi @jikunshang could you help to verify if this condition applies to XPU correctly?

Comment (jikunshang) [reply vllm/model_executor/kernels/linear/scaled_mm/xpu.py]: cc @zufangzhu @Yejing-Lai

Comment (zufangzhu) [inline vllm/model_executor/kernels/linear/scaled_mm/xpu.py]: @fxmarty-amd  please remove this condition.

**Review by zufangzhu — CHANGES_REQUESTED** [review]

File: vllm/model_executor/kernels/linear/scaled_mm/xpu.py (hunk 2 above)
Comment (fxmarty-amd) [reply vllm/model_executor/kernels/linear/scaled_mm/xpu.py]: Right, thanks https://github.com/vllm-project/vllm-xpu-kernels/pull/235

Comment (fxmarty-amd) [reply vllm/model_executor/kernels/linear/scaled_mm/xpu.py]: See below

File: tests/kernels/quantization/test_fp8_kernel_selection.py (hunk 1 above)
Comment (fxmarty-amd) [reply tests/kernels/quantization/test_fp8_kernel_selection.py:63]: Added in ab71aa878f6f5c6df34ca7720c622610ddceb767, although `BlockWiseTorchFP8ScaledMMLinearKernel` is not part of:

https://github.com/vllm-project/vllm/blob/342b8ebd8bd4595826f29ff95dfc48679a03a95a/vllm/model_executor/kernels/linear/__init__.py#L443-L446

File: vllm/model_executor/layers/quantization/fp8.py [hunk 3]
```diff
@@ -278,9 +279,16 @@ def __init__(self, quant_config: Fp8Config):
                 static=self.act_q_static,
                 group_shape=GroupShape(1, self.weight_block_size[0]),
             )
-            self.weight_quant_key = create_fp8_quant_key(
-                static=True, group_shape=GroupShape(*self.weight_block_size)
-            )
+            if self.is_scale_e8m0 and self.weight_block_size == [128, 128]:
+                self.weight_quant_key = kFp8Static128BlockE8M0Sym
+            else:
+                self.weight_quant_key = create_fp8_quant_key(
```
Comment (BowenBao) [inline vllm/model_executor/layers/quantization/fp8.py:285]: can `create_fp8_quant_key` return `kFp8Static128BlockE8M0Sym` if condition matches?

**Review by BowenBao — APPROVED** [review]
LGTM, reasonable improvement on clarification.
